File: src/main/java/matsya/config/ClusterConfig.java
package matsya.config;

import com.google.common.collect.Maps;
import com.typesafe.config.Config;
import com.typesafe.config.ConfigValue;

import java.util.Map;

public class ClusterConfig {
    private String spotASG;
    private String odASG;
    private String machineType;
    // FIXME - Add support for non-VPC based setup as well
    /**
     * Key -> AZ
     * Value -> Subnet
     */
    private Map<String, String> subnets;
    private double threshold;
    private int nrOfTimes;
    private double bidPrice;
    private boolean fallbackToOnDemand;

    /* default */ static ClusterConfig from(Config config) {
        Config subnetConfigs = config.getConfig("subnets");
        Map<String, String> subnets = Maps.newHashMap();
        for (Map.Entry<String, ConfigValue> azAndSubnet : subnetConfigs.entrySet()) {
            subnets.put(azAndSubnet.getKey(), azAndSubnet.getValue().unwrapped().toString());
        }

        return new ClusterConfig()
                .setSpotASG(config.getString("spot-asg"))
                .setOnDemandASG(config.getString("od-asg"))
                .setMachineType(config.getString("machine-type"))
                .setBidPrice(config.getDouble("bid-price"))
                .setSubnets(subnets)
                .setThreshold(config.getDouble("threshold"))
                .setFallbackToOnDemand(config.getBoolean("fallback-to-od"))
                .setNrOfTimes(config.getInt("nr-of-times"));
    }

    public String getSpotASG() {
        return spotASG;
    }

    public ClusterConfig setSpotASG(String spotASG) {
        this.spotASG = spotASG;
        return this;
    }

    public String getMachineType() {
        return machineType;
    }

    public ClusterConfig setMachineType(String machineType) {
        this.machineType = machineType;
        return this;
    }

    public Map<String, String> getSubnets() {
        return subnets;
    }

    public ClusterConfig setSubnets(Map<String, String> subnets) {
        this.subnets = subnets;
        return this;
    }

    public double getThreshold() {
        return threshold;
    }

    public ClusterConfig setThreshold(double threshold) {
        this.threshold = threshold;
        return this;
    }

    public int getNrOfTimes() {
        return nrOfTimes;
    }

    public ClusterConfig setNrOfTimes(int nrOfTimes) {
        this.nrOfTimes = nrOfTimes;
        return this;
    }

    public double getBidPrice() {
        return bidPrice;
    }

    public ClusterConfig setBidPrice(double bidPrice) {
        this.bidPrice = bidPrice;
        return this;
    }

    public boolean isFallbackToOnDemand() {
        return fallbackToOnDemand;
    }

    public ClusterConfig setFallbackToOnDemand(boolean fallbackToOnDemand) {
        this.fallbackToOnDemand = fallbackToOnDemand;
        return this;
    }

    public String getOdASG() {
        return odASG;
    }

    public ClusterConfig setOnDemandASG(String odASG) {
        this.odASG = odASG;
        return this;
    }
}

File: src/main/scala/in/ashwanthkumar/matsya/ClusterConfig.scala
package in.ashwanthkumar.matsya

import java.io.File
import java.util.concurrent.TimeUnit

import com.typesafe.config.{Config, ConfigFactory}
import in.ashwanthkumar.matsya.spotfleet.SpotFleetClusterConfig

import scala.collection.mutable

case class ClusterConfig(name: String,
                         spotASG: String,
                         odASG: String,
                         machineType: String,

                         /**
                          * Key -> AZ
                          * Value -> Subnet
                          */
                         subnets: Map[String, String],
                         maxThreshold: Double,
                         maxNrOfTimes: Int, // TODO - Should we use duration insted of nr of times? 
                         maxBidPrice: Double,
                         odPrice: Double,
                         fallBackToOnDemand: Boolean,
                         odCoolOffPeriodInMillis: Long) {

  def allAZs = subnets.keySet
  def getMachineType = this.machineType
}

object ClusterConfig {
  def from(config: Config): ClusterConfig = {
    val subnetConfigs = config.getConfig("subnets")
    import scala.collection.JavaConversions._
    val subnets = subnetConfigs.entrySet.map(entry => (entry.getKey, entry.getValue.unwrapped().toString)).toMap
    val fallBackToOD = config.getBoolean("fallback-to-od")
    val coolOffDuration = if(fallBackToOD) config.getDuration("od-cool-off-period", TimeUnit.MILLISECONDS) else Long.MaxValue

    ClusterConfig(
      name = config.getString("name"),
      spotASG = config.getString("spot-asg"),
      odASG = config.getString("od-asg"),
      machineType = config.getString("machine-type"),
      subnets = subnets,
      maxThreshold = config.getDouble("max-threshold"),
      maxNrOfTimes = config.getInt("nr-of-times"),
      maxBidPrice = config.getDouble("bid-price"),
      odPrice = config.getDouble("od-price"),
      fallBackToOnDemand = fallBackToOD,
      odCoolOffPeriodInMillis = coolOffDuration
    )
  }
}

case class MatsyaConfig(asgClusters: List[ClusterConfig] = List(), spotFleetClusters: List[SpotFleetClusterConfig] = List(),
                        workingDir: String, slackWebHook: Option[String]) {
  def stateDir = workingDir + "/" + "state"
  def historyDir = workingDir + "/" + "history"
  def getRegion = "us-east-1"
  //FIXME what should the machine types be for spotFleet?
  def machineTypes = asgClusters.map(_.machineType).toSet //++ spotFleetClusters.map(_.getMachineType)
}

object MatsyaConfig {
  def from(config: Config) = {
    import scala.collection.JavaConverters._
    val asgClustersConfig = if(config.hasPath("asgClusters")) config.getConfigList("asgClusters").asScala.toList
    else List.empty[Config]
    val spotFleetClustersConfig = if(config.hasPath("spotFleetClusters")) config.getConfigList("spotFleetClusters").asScala.toList
    else List()
    val asgClusters = asgClustersConfig.map(ClusterConfig.from)
    val spotFleetClusters = spotFleetClustersConfig.map(SpotFleetClusterConfig.from)

    MatsyaConfig(
      asgClusters = asgClusters,
      spotFleetClusters = spotFleetClusters,
      workingDir = config.getString("working-dir"),
      slackWebHook = optionString(config, "slack-webhook")
    )
  }

  private def optionString(config: Config, key: String) = {
    if (config.hasPath(key)) Some(config.getString(key))
    else None
  }
}

object ConfigReader {
  val NAMESPACE = "matsya"

  def readFrom(file: String): MatsyaConfig = load(ConfigFactory.parseFile(new File(file)))

  def load(name: String): MatsyaConfig = load(ConfigFactory.load(name))

  def load(config: Config): MatsyaConfig = {
    val globalConfig: Config = config.getConfig(NAMESPACE)
    MatsyaConfig.from(globalConfig)
  }
}